### PR TITLE
Improve conformance of subresources.kubvirt.io openapi spec

### PR DESCRIPTION
**Pull request by dollierp**
**What this PR does / why we need it**:
Due to non conformant OpenAPI definition, [ZAP](https://www.zaproxy.org/) scanner aborts when trying to discover the KubeVirt subresources API.

This patch series fixes the issues reported by ZAP so that the security scan can proceed.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes https://issues.redhat.com/browse/CNV-34685

**Special notes for your reviewer**:

**Checklist**

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/virt-api/api.go [hunk 1]
```diff
@@ -229,6 +230,7 @@ func (app *virtAPIApp) composeSubresources() {
 
 		restartRouteBuilder := subws.PUT(definitions.NamespacedResourcePath(subresourcesvmGVR)+definitions.SubResourcePath("restart")).
 			To(subresourceApp.RestartVMRequestHandler).
+			Consumes(mime.MIME_ANY).
```
Comment (xpivarc) [inline pkg/virt-api/api.go:233]: Shouldn't this be only json?

Comment (dollierp) [reply pkg/virt-api/api.go:233]: I tried using `restful.MIME_JSON` but the server replies `415: Unsupported Media Type`.

I suspect it is due to `virtctl` not sending any `Content-Type` header.

Comment (fossedihelm) [reply pkg/virt-api/api.go:233]: Quite strange, generally we set the ContentType:
https://github.com/kubevirt/kubevirt/blob/main/staging/src/kubevirt.io/client-go/kubecli/kubecli.go#L327-L332
I'd expect the problem to be only in some stream requests as `virtctl console`...

Comment (dollierp) [reply pkg/virt-api/api.go:233]: Unfortunately it happens also with (at least) start operation.

```
> virtctl start cirros-container-disk

PUT /apis/subresources.kubevirt.io/v1/namespaces/default/virtualmachines/cirros-container-disk/start HTTP/1.1\r\n
Host: 127.0.0.1:8001\r\n
User-Agent: Go-http-client/1.1\r\n
Content-Length: 2\r\n
Accept: application/json, */*\r\n
Accept-Encoding: gzip\r\n
\r\n
{}

HTTP/1.1 415 Unsupported Media Type\r\n
Audit-Id: 9d709cc9-3d60-4c0d-a811-c1e2d17f9c38\r\n
Cache-Control: no-cache, private\r\n
Content-Length: 27\r\n
Content-Type: text/plain; charset=utf-8\r\n
Date: Fri, 17 Nov 2023 10:10:09 GMT\r\n
Strict-Transport-Security: max-age=31536000; includeSubDomains; preload\r\n
X-Kubernetes-Pf-Flowschema-Uid: 55d2a2f8-a04f-465f-98db-5f9fd7b0b4d5\r\n
X-Kubernetes-Pf-Prioritylevel-Uid: 9fd36ad4-180c-4076-8853-0aced74c1420\r\n
\r\n
415: Unsupported Media Type
```

Comment (fossedihelm) [reply pkg/virt-api/api.go:233]: :thinking: will have a look

Comment (fossedihelm) [reply pkg/virt-api/api.go:233]: > https://github.com/kubevirt/kubevirt/blob/main/staging/src/kubevirt.io/client-go/kubecli/kubecli.go#L327-L332 

Looks like this is used to set the `Accept` header and not the `Content-type`

**Comment by fossedihelm** [comment]
/lgtm
@xpivarc can you have a look when you have some spare time? Thanks

File: pkg/virt-api/api.go [hunk 2]
```diff
@@ -386,6 +386,7 @@ func (app *virtAPIApp) composeSubresources() {
 
 		subws.Route(subws.PUT(definitions.NamespacedResourceBasePath(expandvmspecGVR)).
 			To(subresourceApp.ExpandSpecRequestHandler).
+			Param(definitions.NamespaceParam(subws)).
```
[the remark this answers is not in the document]
Comment (dollierp) [reply pkg/virt-api/api.go:397]: There is no `name` param in the URL path of this method in the current OpenAPI definition:

```
/apis/subresources.kubevirt.io/v1/namespaces/{namespace}/expand-vm-spec
```

kubecli code seems to agree with this: https://github.com/kubevirt/kubevirt/blob/main/staging/src/kubevirt.io/client-go/kubecli/instancetype.go#L27.

**Comment by dollierp** [comment]
Hi @fossedihelm / @xpivarc,

Would it be possible to get this one backported to branch release-1.1?

Thanks,

**Pull request merged**